### ISSUE TITLE
Feature: `menu!` macro allowing to add submenus

### ISSUE THREAD

**Issue by jirijakes** [Relm4/Relm4#400]
The `menu!` macro for creating menus is awesome, however at the moment it seems not to support creating submenus.

It might look like this, for example:

```rust
menu! {
  main_menu: {
    "Item 1" => Action1,
    "Submenu": {
      "Subitem 1" => Subaction1
    },
    "Item 2" => Action2
  }
}
```

If it looks OK, please assign the issue to me, I will try to add it. Judging by a quick look, it may not be too difficult.

**Comment by AaronErhardt** [comment]
Sry, I didn't read the whole message and already had a quick shot at the implementation :sweat_smile:

I've published my changes on the [nested-menus](https://github.com/Relm4/Relm4/tree/nested-menus) branch. The code is function but still needs to be cleaned up. Feel free to continue the work or come up with your own solution.

I think the colon isn't necessary, it's better to just use `"Submenu" { items... }`.

**Comment by jirijakes** [comment]
Thanks, Aaron :-D Yeah, I would like to take it over, if that's OK. Looks like a good issue for practice.

**Comment by jirijakes** [comment]
Aaron, I tried to play with it using a bit different approach:

```diff
enum MenuItem {
    Entry(Box<MenuEntry>),
    Custom(LitStr),
    Section(MenuSection),
+   Submenu(Box<Submenu>),
}
```

But that turned out not to work well. I guess thanks to your experience you understand why. So I think what you used in your branch is the correct way.

The code in `nested-menus` doesn't seem to need more work. I would not want to take credit for your work, so please, go ahead and merge your changes.

Thanks, Aaron! At the end, it was a good macro experience to me.

**Comment by AaronErhardt** [comment]
Thanks for your effort regardless! I've just opened a PR, feel free to leave a review.

The problem with your approach is probably that you can't look ahead far enough to figure out what comes after the expression (which can have arbitrary lengths). This is why I have opted for parsing the expression first and then look whether it is an action or a submenu.

**Comment by jirijakes** [comment]
Yeah, it was not so straightforward to distinguish between `Submenu` and `Entry` on the `MenuItem` level. Because I could not, as you say, easily see what comes after the first expression. I guess someone who works with macros all the time could see it right away, I could not and had to find out by trying :-)
